File: StrukturDataBST/BST.java
import java.io.*;
import java.util.*;

class Node {
    int id;
    String nama;
    Node left, right;

    Node(int id, String nama) {
        this.id = id;
        this.nama = nama;
        left = right = null;
    }
}

public class BST {

    Node root;
    int count = 0;

    // ================= INSERT =================
    Node insert(Node root, int id, String nama) {
        if (root == null) {
            count++;
            return new Node(id, nama);
        }

        if (id < root.id)
            root.left = insert(root.left, id, nama);
        else if (id > root.id)
            root.right = insert(root.right, id, nama);
        else
            System.out.println("ID sudah ada!");

        return root;
    }

    // ================= SEARCH =================
    Node search(Node root, int id) {
        if (root == null || root.id == id)
            return root;

        if (id < root.id)
            return search(root.left, id);

        return search(root.right, id);
    }

    // ================= DELETE =================
    Node delete(Node root, int id) {
        if (root == null) return root;

        if (id < root.id)
            root.left = delete(root.left, id);
        else if (id > root.id)
            root.right = delete(root.right, id);
        else {
            // 0 / 1 child
            if (root.left == null) {
                count--;
                return root.right;
            } else if (root.right == null) {
                count--;
                return root.left;
            }

            // 2 children
            Node temp = minValue(root.right);
            root.id = temp.id;
            root.nama = temp.nama;
            root.right = delete(root.right, temp.id);
        }
        return root;
    }

    Node minValue(Node node) {
        while (node.left != null)
            node = node.left;
        return node;
    }

    // ================= TRAVERSAL =================
    void inorder(Node root) {
        if (root != null) {
            inorder(root.left);
            System.out.println(root.id + " - " + root.nama);
            inorder(root.right);
        }
    }

    void preorder(Node root) {
        if (root != null) {
            System.out.println(root.id + " - " + root.nama);
            preorder(root.left);
            preorder(root.right);
        }
    }

    void postorder(Node root) {
        if (root != null) {
            postorder(root.left);
            postorder(root.right);
            System.out.println(root.id + " - " + root.nama);
        }
    }

    // ================= DISPLAY =================
    void showAll() {
        if (root == null) {
            System.out.println("Tree kosong");
            return;
        }
        System.out.println("\n=== DATA (INORDER) ===");
        inorder(root);
    }

    // ================= HEIGHT =================
    int height(Node root) {
        if (root == null) return 0;
        return 1 + Math.max(height(root.left), height(root.right));
    }

    // ================= STAT =================
    void showStat() {
        System.out.println("\n=== INFO TREE ===");
        System.out.println("Jumlah data : " + count);
        System.out.println("Tinggi tree : " + height(root));
    }

    // ================= LOAD CSV =================
    void loadCSV(String filename) {
        try {
            BufferedReader br = new BufferedReader(new FileReader(filename));
            String line;

            br.readLine(); // skip header

            int sukses = 0;

            while ((line = br.readLine()) != null) {
                String[] data = line.split(",");
                int id = Integer.parseInt(data[0].trim());
                String nama = data[1].trim();

                root = insert(root, id, nama);
                sukses++;
            }

            br.close();
            System.out.println("Berhasil load " + sukses + " data!");

        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
    }

    // ================= MAIN =================
    public static void main(String[] args) {
        BST tree = new BST();
        Scanner sc = new Scanner(System.in);

        while (true) {
            System.out.println("\n===== MENU BST =====");
            System.out.println("1. Tambah Data");
            System.out.println("2. Cari Data");
            System.out.println("3. Hapus Data");
            System.out.println("4. Inorder");
            System.out.println("5. Preorder");
            System.out.println("6. Postorder");
            System.out.println("7. Tampilkan Semua Data");
            System.out.println("8. Info Tree");
            System.out.println("9. Load dari CSV");
            System.out.println("0. Exit");

            System.out.print("Pilih: ");
            int pilih = sc.nextInt();
            sc.nextLine();

            switch (pilih) {
                case 1:
                    System.out.print("ID: ");
                    int id = sc.nextInt();
                    sc.nextLine();
                    System.out.print("Nama: ");
                    String nama = sc.nextLine();
                    tree.root = tree.insert(tree.root, id, nama);
                    break;

                case 2:
                    System.out.print("Cari ID: ");
                    int cari = sc.nextInt();
                    Node hasil = tree.search(tree.root, cari);
                    if (hasil != null)
                        System.out.println("Ditemukan: " + hasil.id + " - " + hasil.nama);
                    else
                        System.out.println("Tidak ditemukan");
                    break;

                case 3:
                    System.out.print("Hapus ID: ");
                    int hapus = sc.nextInt();
                    tree.root = tree.delete(tree.root, hapus);
                    break;

                case 4:
                    tree.inorder(tree.root);
                    break;

                case 5:
                    tree.preorder(tree.root);
                    break;

                case 6:
                    tree.postorder(tree.root);
                    break;

                case 7:
                    tree.showAll();
                    break;

                case 8:
                    tree.showStat();
                    break;

                case 9:
                    System.out.print("Nama file CSV: ");
                    String file = sc.nextLine();
                    tree.loadCSV(file);
                    break;

                case 0:
                    System.out.println("Program selesai");
                    return;

                default:
                    System.out.println("Menu tidak tersedia");
            }
        }
    }
}

File: StrukturDataBST/BST.py
import csv

class Node:
    def __init__(self, id, nama):
        self.id = id
        self.nama = nama
        self.left = None
        self.right = None

class BST:
    def __init__(self):
        self.root = None
        self.count = 0

    # INSERT
    def insert(self, root, id, nama):
        if root is None:
            self.count += 1
            return Node(id, nama)

        if id < root.id:
            root.left = self.insert(root.left, id, nama)
        elif id > root.id:
            root.right = self.insert(root.right, id, nama)

        return root

    # SEARCH
    def search(self, root, id):
        if root is None:
            return None
        if id == root.id:
            return root
        elif id < root.id:
            return self.search(root.left, id)
        else:
            return self.search(root.right, id)

    # DELETE
    def delete(self, root, id):
        if root is None:
            return root

        if id < root.id:
            root.left = self.delete(root.left, id)
        elif id > root.id:
            root.right = self.delete(root.right, id)
        else:
            # 1 child / no child
            if root.left is None:
                self.count -= 1
                return root.right
            elif root.right is None:
                self.count -= 1
                return root.left

            # 2 children
            temp = self.minValue(root.right)
            root.id = temp.id
            root.nama = temp.nama
            root.right = self.delete(root.right, temp.id)

        return root

    def minValue(self, node):
        current = node
        while current.left:
            current = current.left
        return current

    # TRAVERSAL
    def inorder(self, root):
        if root:
            self.inorder(root.left)
            print(root.id, "-", root.nama)
            self.inorder(root.right)

    def preorder(self, root):
        if root:
            print(root.id, "-", root.nama)
            self.preorder(root.left)
            self.preorder(root.right)

    def postorder(self, root):
        if root:
            self.postorder(root.left)
            self.postorder(root.right)
            print(root.id, "-", root.nama)

    # LOAD CSV
    def load_csv(self, filename):
        try:
            with open(filename, newline='', encoding='utf-8') as file:
                reader = csv.DictReader(file)

                for row in reader:
                    id = int(row["ID"])
                    nama = row["Nama"]
                    self.root = self.insert(self.root, id, nama)

            print("Data berhasil di-load dari CSV!")

        except FileNotFoundError:
            print("File tidak ditemukan!")
        except Exception as e:
            print("Error:", e)


# MAIN PROGRAM
bst = BST()

while True:
    print("\n=== MENU BST ===")
    print("1. Insert Data")
    print("2. Search Data")
    print("3. Delete Data")
    print("4. Inorder")
    print("5. Preorder")
    print("6. Postorder")
    print("7. Load dari CSV")
    print("8. Exit")

    pilih = input("Pilih: ")

    if pilih == "1":
        id = int(input("ID: "))
        nama = input("Nama: ")
        bst.root = bst.insert(bst.root, id, nama)

    elif pilih == "2":
        id = int(input("Cari ID: "))
        hasil = bst.search(bst.root, id)
        if hasil:
            print("Ditemukan:", hasil.id, "-", hasil.nama)
        else:
            print("Tidak ditemukan")

    elif pilih == "3":
        id = int(input("Hapus ID: "))
        bst.root = bst.delete(bst.root, id)

    elif pilih == "4":
        bst.inorder(bst.root)

    elif pilih == "5":
        bst.preorder(bst.root)

    elif pilih == "6":
        bst.postorder(bst.root)

    elif pilih == "7":
        file = input("Nama file CSV: ")
        bst.load_csv(file)

    elif pilih == "8":
        break